pmm-app/src/pmm-qan-app-panel/QueryAnalytics/StateContext.tsx

```diff
@@ -0,0 +1,279 @@
+        value: state.labels[filter] || ['All'],
+      };
+    });
+    templateVariables[0].variableSrv.variableUpdated(templateVariables[0]);
+  }
+
+  static generateURL(state) {
+    // read parameters and create new url
+    // @ts-ignore
+    const labels =
+      state.labels &&
+      Object.keys(state.labels)
+        .map(key => {
+          // @ts-ignore
+          const variables = state.labels[key];
+          return variables.map(variable => `var-${key}=${variable}`).join('&');
+        })
+        .filter(Boolean)
+        .join('&');
+    const columnsQuery = state.columns ? `columns=${JSON.stringify(state.columns)}` : '';
+    const groupBy = state.groupBy ? `group_by=${state.groupBy}` : '';
+    const filterByQuery = state.queryId ? `filter_by=${state.queryId}` : '';
+    const orderBy = state.orderBy ? `order_by=${state.orderBy}` : '';
+    const from = state.rawTime && state.rawTime.from ? `from=${state.rawTime.from}` : '';
+    const to = state.rawTime && state.rawTime.to ? `to=${state.rawTime.to}` : '';
+    // TODO: replace crutch with right redirect
+    return `${window.location.pathname}?${[columnsQuery, filterByQuery, labels, orderBy, groupBy, from, to]
+      .filter(Boolean)
+      .join('&')}`;
+  }
+
+  static parseURL(query) {
+    const urlParams = {} as any;
+    urlParams.from = ParseQueryParamDate.transform(query.get('from') || 'now-12h', 'from')
+      .utc()
+      .format('YYYY-MM-DDTHH:mm:ssZ');
+    urlParams.to = ParseQueryParamDate.transform(query.get('to') || 'now', 'to')
+      .utc()
+      .format('YYYY-MM-DDTHH:mm:ssZ');
+    // urlParams.rawTime = {
+    //   from: query.get('from'),
+    //   to: query.get('to'),
+    // };
+    urlParams.columns = JSON.parse(query.get('columns')) || DEFAULT_COLUMNS;
+    urlParams.labels = ContextActions.setFilters(query);
+    urlParams.pageNumber = 1;
+    urlParams.pageSize = 10;
+    urlParams.orderBy = query.get('order_by') || `-${urlParams.columns[0]}`;
+    urlParams.queryId = query.get('filter_by');
+    urlParams.querySelected = !!query.get('filter_by');
+    urlParams.groupBy = query.get('group_by') || 'queryid';
+    // console.log(Object.assign({}, urlParams));
+    return urlParams;
+  }
+
+  static setLabels(filters) {
+    const labels = {};
+    Object.keys(filters)
+      .filter(filter => filters[filter])
+      .forEach(filter => {
+        const [group, value] = filter.split(':');
+        // TODO: using '--' because final form think that it is a nested fields
+        //  need to replace it with something better
+        if (labels[group]) {
+          labels[group].push(value.replace(/\-\-/gi, '.'));
+        } else {
+          labels[group] = [value.replace(/\-\-/gi, '.')];
+        }
+      });
+    return labels;
+  }
+}
+
+const actions = {
+  setLabels: value => state => {
+    const newState = { ...state, labels: ContextActions.setLabels(value), pageNumber: 1 };
+    delete newState.queryId;
```

Comment:
Pls use `omit`, since `delete` is not that reliable should the property be found in the prototype :)